### PR TITLE
Deprecate offset_position="data".

### PR DESCRIPTION
**Pull request by anntzer**
The `offset_position` property of collections is currently set to "data"
only for hexbin(), but one can do away with it by just setting the
correct offsetTrans instead -- which this PR does.  The advantage of
doing so is that the correct offsetTrans only needs to be implemented
once, whereas support for `offset_position` needs to be implemented for
each renderer class separately (including at the C-level for the Agg
backend).

Note that the *offset_position* argument to `draw_path_collection` is
not pending removal because its removal would require coordination with
third-party backends; the plan is to just always set it to "screen"
after the deprecation period.

If there is demand for it, _VectorTransform could be made a public API.

---

TLDR: hexbin() currently (effectively) uses its own code-path in renderer code, but that's not needed.

Edit: closes https://github.com/matplotlib/matplotlib/issues/16461 and (effectively) https://github.com/matplotlib/matplotlib/issues/16496

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/transforms.py [hunk 1]
```diff
@@ -2680,6 +2680,28 @@ def get_matrix(self):
...
+
+    def __init__(self, transform, **kwargs):
+        super().__init__(**kwargs)
+        self._base_transform = transform
+
+    def __str__(self):
+        return ("{}(\n"
+                    "{})"
+                .format(type(self).__name__,
+                        _indent_str(self._base_transform)))
```
Comment (ImportanceOfBeingErnest) [inline lib/matplotlib/transforms.py]: Is there a reason for this 4-line construct? Would this not be the same as 

        return "{}(\n{})".format(type(self).__name__, 
                                  _indent_str(self._base_transform))

**Comment by ImportanceOfBeingErnest** [comment]
In general, this makes sense to me. 

Is there anything that would speak against making `VectorTransform` public? That would allow to directly let users know what to use in lieu of the deprecated `"data"` argument, without the need to reinterate the complete thing in case someone does have a custom use case of it.

File: lib/matplotlib/transforms.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/transforms.py]: This is consistent with how all other reprs are defined in this module (i.e. the format string's layout matches the final result).

**Comment by anntzer** [comment]
I would strongly doubt that there are many users of offset_position="data".  But if you're worried about this we can replace the Deprecation by a PendingDeprecation, as usual.  (On the other hand, seeing whether we get complaints or not is also one of the points of the deprecation period.)
At least I haven't fully thought about what the public API re: VectorTransform should be, and I would perhaps like to use something similar to replace some of the annotation coord strings (cf discussion at #13515).

**Comment by ImportanceOfBeingErnest** [comment]
Even if the amount of users hit by this isn't massive, it's still not the right policy to first break something and then wait if someone complains. Often enough, people might just be frustrated without opening an issue about it. In this case I think that  the functionality of an offset transform in data coordinates is in general useful. I would agree that this does not require an additional argument, but should be incorporated into the transform itself. So removing `offset_position` in the long run makes sense. But if no alternative can be provided by now, the whole thing can be postponed up to when a general concept for such offset transforms is developed.

**Comment by anntzer** [comment]
The changelog entry can mention e.g. "if you need this functionality, you can vendor the implementation of _VectorTransform" (which is quite short if you don't bother with `__str__`) "Please let us know if that's the case; then we will consider making this a public API".  (If one is really using offset_transform="data", that's a fairly low level thing and vendoring the transform code is really not that hard once you already have to update the call sites anyways.)

I simply disagree that we can never deprecate things on the basis that someone may possibly be using them; as always there is a cost-benefit ratio that needs to be considered, and certainly if someone actually complains about it we'll consider it more.

**Comment by anntzer** [comment]
(Some notes mostly to self: there's already transforms.offset_copy and text.OffsetFrom on the same theme.)

**Comment by efiring** [comment]
@anntzer If I understand correctly, hexbin could be implemented without `offset_position` *or* the new `_VectorTransform`.  Would such an implementation be inferior?  Slower, for example?
The transforms machinery has always been hard to understand, visualize, and remember, at least for me, and it would be good if any changes that are made serve to make it a little less difficult, a little more readable.  The name `_VectorTransform` is not helping.

**Comment by anntzer** [comment]
The problem is that without _VectorTransform (I'm not particularly in love with the name), collections (or more specifically, _CollectionWithSizes, which PolyCollection (the class returned by hexbin()) inherits) will by default size their components (the hexagons, in this case) to have a size in physical units (points), whereas we want to have them in data units, which is what _VectorTransform achieves.  (Previously, this had to be built in to every renderer class, _VectorTransform just factors out that functionality.)

We could instead use PatchCollection, which just takes a bunch of patches and plots them as is (see e.g. the very recent thread on the mailing list https://mail.python.org/pipermail/matplotlib-users/2019-May/001663.html); however such an approach would *at least* lead to more inefficient vector output because the vector file would need to include every hexagon on every point, rather than a single hexagon and encode a list of offsets.  In fact, this is the very reason why offset_position exists (#859); that PR indicates that just dropping the amount of data sent from python to agg also significantly accelerates agg.

**Comment by tacaswell** [comment]
Pushing out to 3.3 as I think this needs to be understood a bit better still.

It also looks like we currently have the ability to cycle through data transfroms (?!) and there is no documented path for a user who was directly creating the collection and passing `data` to adjust their code.

File: lib/matplotlib/transforms.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/transforms.py]: this was rewritten using _make_str_method.

**Comment by ImportanceOfBeingErnest** [comment]
I would like to reiterate the possibility to make `_VectorTransform` public. There has just been #16548, where of course one could draw circles in data coordinates, but in addition to the options shown in the [SO post](https://stackoverflow.com/questions/48172928/scale-matplotlib-pyplot-axes-scatter-markersize-by-x-scale/48174228#48174228) OP linked to, probably the fastest option would be to use a `PathCollection`. 

```
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.collections as mcoll
import matplotlib.transforms as mtrans
import matplotlib.markers as mmarkers

c = mcoll.PathCollection(
                [mmarkers.MarkerStyle("o").get_path()],
                offsets=np.repeat(np.arange(10),2).reshape(10,2),
                edgecolors=["k"]*10,
                transOffset=mtrans.IdentityTransform(),
                offset_position="data")
plt.gca().add_collection(c)
plt.gca().set(xlim=(-2, 11), ylim=(-2,11))
plt.gca().set_aspect("equal")
plt.show()
```

That option is gone if `offset_position="data"` is deprecated without replacement.

**Comment by anntzer** [comment]
OK, I made VectorTransform public, but marked it as experimental to allow for later changes depending on actual use..

File: lib/matplotlib/collections.py [hunk 2]
```diff
@@ -407,7 +410,7 @@ def contains(self, mouseevent):
                self._picker is not True  # the bool, not just nonzero or 1
             else self._pickradius)
 
-        if self.axes and self.get_offset_position() == "data":
+        if self.axes:
```
Comment (tacaswell) [inline lib/matplotlib/collections.py]: This will un-stale the view limits more often, is that OK?

Comment (anntzer) [reply lib/matplotlib/collections.py]: Properly fixing that would be something like
```
if self.axes and self.get_transform().contains_branch(self.axes.transData)
```
but this currently returns False because VectorTransform doesn't "contain" transData (i.e. `VectorTransform(transData) - transData` can't be simplified out) -- really we'd need another relationship of "depends_on(transData)" -- or we can just declare that that's what contains_branch means.  (This kind of annoying details is another reason I was not so keen on making VectorTransform public, but heh.)

OTOH it's just a minor perf loss (the unstaling would only be done once) and it's not worse than before the lazyfication of viewLims, so I don't think it's too bad either.

Comment (tacaswell) [reply lib/matplotlib/collections.py]: Good enough.

File: lib/matplotlib/transforms.py [hunk 3]
```diff
@@ -2612,6 +2612,26 @@ def get_matrix(self):
         return self._mtx
 
 
+class VectorTransform(Affine2DBase):
```
Comment (timhoffm) [inline lib/matplotlib/transforms.py]: I‘m not quite happy with the name. „Vector“ has a lot of semantics, but IMHO only a loose connection to „no offset“. Some possible alternatives:

- `WithoutOffsetTransform`
- `RotationOnlyTransform`
- `IgnoreOffsetTransfrom`

Possibly even with an `...Wrapper` suffix to indicate that this modifies / builds upon another transform.

Comment (anntzer) [reply lib/matplotlib/transforms.py]: I picked vector as it's meant to transform the delta vector between a pair of points, rather than individual points (I agree this may not be obvious). Ignore/WithoutOffset seems to hide the relevant semantics to me and focus more on how they are achieved, and RotationOnly is wrong as there can also be stretch.
I'm open to changing the name, though.

**Comment by timhoffm** [comment]
> Ignore/WithoutOffset seems to hide the relevant semantics to me and focus more on how they are achieved

Can you paraphrase that semantics (maybe even directly in the docstring)? I think I probably haven't fully understood `offset_position='data'` and `VectorTransform`, but maybe that's also the point. The current docstring just says

> A transform wrapper that ignores the offset component of a base transform.

And that's what I've been trying to pick up for the name. That description and the name `VectorTransform` do not make a connection in my head.

**Comment by anntzer** [comment]
Edited the docstring, does this look better?

**Comment by timhoffm** [comment]
Thanks, the extended description looks good. I still feel a mismatch between the class name and the summary line. IMHO they should be on the same level of abstraction. So either

~~~
class VectorTransform(Affine2DBase):
    """
    A transform for connection vectors between points.

    This wraps another transform and ignores its offset component.
    ....
~~~
or 
~~~
class IgnoreOffsetTransfrom(Affine2DBase):
    """
    A transform wrapper that ignores the offset component of a base transform.
    ....
~~~

I think that, more precisely, the vectors you think of are direction vectors (in contrast to  position vectors). Maybe `DirectionVectorTransform` is a suitable alternative name?

**Comment by anntzer** [comment]
Ah, I see the confusion now.  Yes, I meant https://en.wikipedia.org/wiki/Displacement_(geometry), not https://en.wikipedia.org/wiki/Position_(geometry).

Pushed a new version.

**Comment by efiring** [comment]
DifferenceTransform? DeltaTransform?  SizeAndShapeTransform?  I'm struggling.  It would really be nice to have a name that would be highly evocative--which for me, VectorTransform still is not. It seems that the zeroing of the offset really is the key to understanding what the wrapper is doing.  So I tend to favor ZeroOffsetTransform, or NoOffsetTransform, or something like that.  Or, if you want to include the mathematical property, it would be DistributiveTransform, since it is modifying a transform to make it distributive with respect to addition.

**Comment by anntzer** [comment]
DisplacementTransform? OffsetTransform?
Or we can just poll this over the call :)

**Comment by timhoffm** [comment]
Let‘s briefly discuss in the call.

**Comment by QuLogic** [comment]
I don't believe this came up in any call; I've put it on the agenda for next week (though we might not get to it.)

**Comment by tacaswell** [comment]
On call we settled on `AffineDeltaTransfrom` for name.

**Comment by anntzer** [comment]
renamed

**Comment by anntzer** [comment]
done, as "position deltas".

File: lib/matplotlib/collections.py [hunk 4]
```diff
@@ -85,6 +85,7 @@ class Collection(artist.Artist, cm.ScalarMappable):
     # subclass-by-subclass basis.
     _edge_default = False
 
+    @cbook._delete_parameter("3.1", "offset_position")
```
Comment (QuLogic) [inline lib/matplotlib/collections.py:88]: 3.3?

Comment (anntzer) [reply lib/matplotlib/collections.py]: oops

**Comment by anntzer** [comment]
rebased

**Comment by QuLogic** [comment]
Travis reporting seems broken; there are not currently any builds running.
